### PR TITLE
JavaScript fallback for WebAssembly libraries

### LINKED ISSUE #6768: Add JS fallback for WASM components
Some of our current crypto algorithms are compiled to WASM (argon2, liboqs). This causes issues in environments where WASM is disabled (browsers on iOS with Lockdown Mode, Tor-browser and derivatives in some security settings).

In the future we would like to compile Rust-based SDK to WASM so we will rely on WASM even more.

We would like to allow those special browser cases to still use Tuta.

## Tasks
- Argon2
   - [x] Try wasm-to-JS transpiler manually for one of our WASM libraries
   - [x] Integrate with build system to produce the fallback
   - [x] Implement WASM detection at runtime
   - [x] Load fallback when needed
 - Liboqs
   - [x] Try wasm-to-JS transpiler manually for one of our WASM libraries
   - [x] Integrate with build system to produce the fallback
   - [x] Implement WASM detection at runtime
   - [x] Load fallback when needed

## Test Notes
See #6835

### PR DESCRIPTION
**Pull request by murilopereirame**
Adds the generation of a JavaScript transpile for Argon2 and Liboqs WebAssembly files. This is done through wasm2js that's builtin on emcc.

Closes #6768

Co-authored-by: wec43 <wec@tutao.de>

### CODE REVIEW

**Comment by charlag** [comment]
Nice!

I think ideally (and I'm not sure if it's in the scope) we wouldn't need to manually take care of every WASM file. In SDK prototype we have successfully used WASM loader:
https://github.com/tutao/tutanota/blob/05b6b826b6b7918e62e926e9d4e4d15f46a7d05c/buildSrc/DevBuild.js#L93

maybe we need another plugin(s) so that we can automatically generate the fallback and add the code that conditionally loads one or another

File: libs/webassembly/Makefile_argon2 [hunk 1]
```diff
@@ -4,12 +4,15 @@
 TOTAL_MEMORY=48MB
 
 CC=emcc
+WASM2JS=wasm2js
```
Comment (charlag) [inline libs/webassembly/Makefile_argon2:7]: I don't think wasm2js should be part of the makefiles. This is one example of it: where does wasm2js come from? If we run it in JS then we can just import it

**Review by charlag — CHANGES_REQUESTED** [review]
great job!

I don't think compiled wasm/js should be checked in though. We should also explicitly depend on wasm2js

File: libs/webassembly/Makefile_argon2 (hunk 1 above)
Comment (murilopereirame) [reply libs/webassembly/Makefile_argon2:7]: wasm2js is inside [`Binaryen`](https://github.com/WebAssembly/binaryen?tab=readme-ov-file) and Binaryen is inside [`Emscripten`](https://github.com/WebAssembly/binaryen/blob/98f08effb18a94618ef12f25134c289836fc316e/README.md?plain=1#L29). So i think it's fine to be part of the make files (?)

Comment (charlag) [reply libs/webassembly/Makefile_argon2:7]: okay yeah that's a good argument

**Comment by charlag** [comment]
We did agree to remove the compiled artifacts, right?

**Comment by charlag** [comment]
Also, for documentation purposes, the build system integration/glue generation is tracked at #6835

File: packages/tutanota-utils/lib/WebAssembly.ts [hunk 2]
```diff
@@ -1,5 +1,36 @@
...
+ */
+type Argon2WebAssemblyFallback = {
+	argon2id_hash_raw: Record<string, WebAssembly.ExportValue>
+}
+
+/*
+ * Functions required to be present inside the Liboqs fallback
+ */
+type KyberWebAssemblyFallback = {
+	OQS_KEM_keypair: Record<string, WebAssembly.ExportValue>
```
Comment (charlag) [inline packages/tutanota-utils/lib/WebAssembly.ts:14]: why is it a `Record`? shouldn't it be `OQS_KEM_KEYPAIR_RawFN`?

**Review by charlag — COMMENTED** [review]
I have another concern, I think right now we will just rollup the fallback JS into the same bundle as the rest of the crypto? it should be separate, maybe it's own chunk

File: packages/tutanota-utils/lib/WebAssembly.ts (hunk 2 above)
Comment (murilopereirame) [reply packages/tutanota-utils/lib/WebAssembly.ts:14]: The idea was giving types here as the same type of WebAssembly so there's not need export these types from tuta crypto library. And yes, it should not be a record and just WebAssembly.ExportValue

**Comment by murilopereirame** [comment]
> We did agree to remove the compiled artifacts, right?

I remember that for this initial implementation we check in the compiled artifacts and after implementing the integration inside the build process we're going to uncheck them.

**Comment by wrdhub** [comment]
pr #6868 has what needs to be merged in, this one can be closed

**Pull request closed without merging**